User profile #19

noelworden opened the pull request.

Commit:
[ FEAT ] Added profile.rb
[ FEAT ] Added profile_spec.rb
[ CHG ] Modified user.rb to assoc with profile.rb
[ FEAT ] Added users and profiles Factories

Started the process of user having a profile. At this point it is only
tested for basic attribute validation

Commit:
[ CHG ] Modified user.rb association with profile.rb

Had the association as `has_many` but that didnt make sense. Changed to
`has_one`

coreyd303 commented:
@noelworden 

In profile.rb, there is no need to validate the length of the first or last name, presence is fine. and we do not need to validate user_id presence, that is handled through active record.

Think carefully about if you do want to validate presence of EVERYthing, it can be a nicer UX to let the user walk away and come back later when completing a form like this.

We don't actually need email on the Profile object, it is already stored on the User object  `user.email`

also, we don't need to declare `has_one :profile` on the user model, this might help clarify 
[has_one vs belongs_to](http://requiremind.com/differences-between-has-one-and-belongs-to-in-ruby-on-rails/)

how do you plan to use the enum in profile.rb? If its just for a form why not a standard constant? is there a purpose to leveraging the enum? Take a look at this:
[why and what is a ruby enum](http://railsguides.net/safe-enums-in-ruby/)

if you think about how to want to access the data, using the enum tactic if I query `user.huntingstyle` what will my implicit return be? Is that returned value actually useable or will I have to do bunch of overhead to figure out which integer is which hunting style? Maybe consider something like

`STYLES = ["guns", "rocks", "something else"]`  this can be called from anywhere in the application (ie a form) as Profile::STYLES, and will then save a simple string to the object. Plus enums in ruby are not immutable, they are essentially a hash so can be injected upon, also they are best when we think something will NEVER change, the nice thing about a constant is that its easy to change in one place and it ripple out. 

The specs look great, you will probably need to tweak them once you make updates, but otherwise really good work!

coreyd303 commented:
OH, but I do like that you are using a regex validation on the email ;) if you wanted to you could drop that validation into the user model, but lets make it more robust, maybe something like
`VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i`

Commit:
[ CHG ] profile.rb to remove unnecessary validations
[ CHG ] user.rb to not have has_one validation
[ CHG ] migrations to change table parameters
[ CHG ] profile_spec to reflect above changes

Commit:
[ REFAC ] user.rb to have huntingstyle constants
[ REFAC ] factories.rb to reflect user constants
[ CHG ] profile_spec to be current

Changed huntingstyles from enums to constants. Proflie factory chooses
a constant at random when building profile. Specs are up to date with
these changes.
